### ISSUE TITLE
Question to task 'config | configuring consul' at retry_join building

### ISSUE THREAD

**Issue by jardleex** [mrlesmithjr/ansible-consul#14]
Hello @mrlesmithjr 

I'm struggling with  at the task _config | configuring consul_ with this error:

```
TASK [mrlesmithjr.consul : config | configuring consul] **************************************************************************************************************************************************************************************
Friday 28 April 2017  11:35:51 +0200 (0:00:00.088)       0:00:14.442 ********** 
fatal: [consul2]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
fatal: [consul3]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
ok: [consul1]

```

I've isolated the issue within this block in _templates/etc/consul.d/config.json.j2 _
```
...
{%   if inventory_hostname != consul_master %}
{%     set _temp_retry_join = [] %}
{%       for host in groups[consul_servers_group] %}
{%         if inventory_hostname != host %}
{%         set _temp_ip = hostvars[host]['consul_bind_address'] %}
{%         set _temp_retry_join = _temp_retry_join.append(_temp_ip) %}
{%         endif %}
{%       endfor %}
{%     set _config = config_json.update({"retry_join": _temp_retry_join}) %}
{%   endif %}
...
```

The node _consul1_ was picked as master by the role, there to task worked fine as no _retry_join_ needs to be set there when I got the official docu right.

I've set this variables via _group_vars_ for my three consul servers:
```
consul_datacenter: "{{ location }}" # custom variable in my env containing a value I can re-use as datacenter
consul_servers_group: "{{ hostgroup }}" # a hostgroup containing all three consul nodes; a variable build out of each FQDN extracting all signs till the first digit so consul1, consul2 and consul3 getting assinged to group consul
consul_enable_dnsmasq: false
consul_services: []
```

I've also another problem with the clients next statement about _consul_clients_group_ in _config.json.j2_. For now I've commented this statement out to bypass the issue. Currently I've no clients to deploy and even overriding _consul_clients_group_ with an empty list _consul_clients_group: []_ does not supress the second error. But for now I'd really appriciate your help on the issue above. 
Maybe I'll open another issue for the later one if the solution found in here may not solve the second issue too.

Thanks in advance!

Jard

**Comment by mrlesmithjr** [comment]
@jardleex Thanks for pointing this out. The latest merge should resolve the issue with the clients you were seeing.

**Comment by jardleex** [comment]
@mrlesmithjr thanks for the quick feedback. Looks better now on the client error.
When I comment the block mentioned above to build the _retry_join_ line now the role completes without errors.

So just the generation of _retry_join_ is an issue still.
Ah, I forgot to note that I commented out the _no_log: true_ line on the _config | configuring consul_ to make the error visible.

Let me know if you need any futher informations!

Best regards

Jard

**Comment by mrlesmithjr** [comment]
I think I may be confused on your retry_join question. I have a 3 node cluster that I just spun up and here is what my `config.json` looks like on each one.

`node0` - master
```
vagrant@node0:/etc/consul.d$ cat config.json
{
    "acl_datacenter": "dc1",
    "acl_default_policy": "allow",
    "acl_down_policy": "extend-cache",
    "acl_master_token": "c9f640c7-7635-4859-9f1a-b96d69265043",
    "bind_addr": "192.168.250.10",
    "bootstrap": true,
    "client_addr": "0.0.0.0",
    "data_dir": "/var/consul",
    "datacenter": "dc1",
    "enable_syslog": true,
    "encrypt": "OtT3CJghJrFeFebjxa+gNw==",
    "log_level": "INFO",
    "node_name": "node0",
    "server": true
}
```
`node1` - Member
```
vagrant@node1:~$ cat /etc/consul.d/config.json
{
    "acl_datacenter": "dc1",
    "acl_default_policy": "allow",
    "acl_down_policy": "extend-cache",
    "acl_master_token": "c9f640c7-7635-4859-9f1a-b96d69265043",
    "bind_addr": "192.168.250.11",
    "bootstrap_expect": 3,
    "client_addr": "0.0.0.0",
    "data_dir": "/var/consul",
    "datacenter": "dc1",
    "enable_syslog": true,
    "encrypt": "OtT3CJghJrFeFebjxa+gNw==",
    "log_level": "INFO",
    "node_name": "node1",
    "retry_join": [
        "192.168.250.10",
        "192.168.250.12"
    ],
    "server": true
}
```
`node2` - Member
```
vagrant@node1:~$ cat /etc/consul.d/config.json
{
    "acl_datacenter": "dc1",
    "acl_default_policy": "allow",
    "acl_down_policy": "extend-cache",
    "acl_master_token": "c9f640c7-7635-4859-9f1a-b96d69265043",
    "bind_addr": "192.168.250.12",
    "bootstrap_expect": 3,
    "client_addr": "0.0.0.0",
    "data_dir": "/var/consul",
    "datacenter": "dc1",
    "enable_syslog": true,
    "encrypt": "OtT3CJghJrFeFebjxa+gNw==",
    "log_level": "INFO",
    "node_name": "node2",
    "retry_join": [
        "192.168.250.10",
        "192.168.250.11"
    ],
    "server": true
}
```

So I am thinking your first issue was actually coming from the issue with the template creation and the consul clients group issue which should be good now. Hope this helps!

**Comment by jardleex** [comment]
Sorry for the confusion, I'll try to clearify:
Your changes in _9cb6a90_ solved my **SECOND** with the clients setup. Initially I was tempted to created a dedicated issue for this but you already solved it.

Now back to the **FIRST** issue.
My role run's are failing when it tries to build the _retry_join_ array in the _config.json.j2_ file. As it works for you out of the box I assume an handling error of the role on my side.
I'm using a dynamic inventory script where it's not that quick'n easy possible to define new groups. As I understood your role I can pass a group name of hosts to the role so it know's who's going to be a server. The variable I used for this is _consul_servers_group_ and gave it my variable _hostgroup_ which includes three VMs (consule1/2/3.something).
When doing so, I'm getting the error posted in the first comment.

I just tried only use the static inventory. Defined a group called _consul_servers_, attached the role to this group and did not defined the var _consul_servers_group_ > same error.

Can nearly feel the facepalm on my forehead when we found the answer...

Best regards

Jard

**Comment by mrlesmithjr** [comment]
@jardleex Man I am sorry, I cannot duplicate your same issue. Here is what I just tried and had no issues.
```
- hosts: consul_all
  # become: true
  vars:
    consul_bind_interface: 'enp0s8'
    consul_datacenter: '{{ location }}'
    consul_servers_group: '{{ hostgroup }}'
    hostgroup: 'jardleex_servers'
    location: 'vagrant'
    pri_domain_name: 'test.vagrant.local'
  roles:
    - role: ansible-consul
      tags:
        - consul
  tasks:
```
inventory:
```
[jardleex_servers]
node0
node1
node2
```

**Comment by jardleex** [comment]
@mrlesmithjr thank's for your reply.

After another three hours I still ain't got a solution. Some things I noticed:
- I assume you are using a static inventory running agains local Vagrant boxes > I'm using a dynamic inventory against remote KVM instances.
- The offending line is ```...{%         set _temp_retry_join = _temp_retry_join.append(_temp_ip) %}...``` complaining that ``fatal: [consul1.mydomain]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}``
- Using a dynamic inventory may return the consul server group members in a different order. This make's the role act as the leader has changed and triggers config changes and service reloads. No big deal, already got an idea how to solve this but ain't directly related to this issue.

As my Python and Jinja2 voodoo is not strong enough I don't get the issue with this. And as currently nobody can duplicate my issue it seems I'm stuck. I already tried Ansible version 2.2 and used another control machine but sill no luck.

Things I tried:
- Downgrade Ansible to 2.2 > no change
- Use static inventory > no change
- manipulate Jinja2 to json loop > only worked when I definined a single server for retry_join with out any .append() but that's not the goal of the loop altered.
- Use other Ansible control machine > no change
- Wiped and rebuild KVM guests from scratch > no change

Some details on my env:
- Ansible version 2.3.0.0
- Python version 2.7.12 on control Ansible host and remote servers
- KVM instances inside Ganeti cluster als Consul nodes
```
# inventory
[jardleex_servers]
consul1.mydomain
consul2.mydomain
consul3.mydomain

# top level playbook
- name: SIT.consul
  hosts: jardleex_servers
  tags: consul
  vars:
    # consul_bind_interface: 'eth0' < no change if set or not
    # pri_domain_name: 'mydomain' < no change if set or not
    consul_servers_group: 'jardleex_servers'
  roles:
    - { role: mrlesmithjr.consul, tags: consul }

```

I'd really like to use this role to deploy Consul at our company and support you on debugging the issue but right now all my ideas are gone. All other Ansible roles for Consul out there (checked ~10) lack the one or another feature, are outdated, aren't idempotent, do more then they should and so on.
Can I provide any furhter informations to demystify this?

Best regards 

Jard

**Comment by mrlesmithjr** [comment]
@jardleex Man that sucks. I will continue to try and duplicate the issue. I have a few things in mind that I think could be causing the issue but will need to mimic the scenario as closely as possible.

**Comment by mrlesmithjr** [comment]
@jardleex Can you try using branch `dev/retry-join-issue#14` and report back if it works correctly now?

**Comment by jardleex** [comment]
@mrlesmithjr just tried out the branch ``dev/retry-join-issue#14`` but still the same error.

Just tried to omit the domain name from the nodes in the static inventory but no change. Here the full debug output of the specifig task. Playbook call was ``./site.yml -i inventory/manual --limit jardleex_servers --tags consul -vvv``

```
TASK [mrlesmithjr.consul : config | configuring consul] **************************************************************************************************************************************************************************************
task path: /home/myusername/Dokumente/git/ansible/galaxy/mrlesmithjr.consul/tasks/config.yml:47
Monday 01 May 2017  19:32:25 +0200 (0:00:00.397)       0:00:09.366 ************ 
<consul1> ESTABLISH SSH CONNECTION FOR USER: ansible
<consul1> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o 'IdentityFile="/home/myusername/.ssh/ansible_rsa"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ansible -o ConnectTimeout=10 -o ControlPath=/home/myusername/.ansible/cp/%h-%r consul1 '/bin/sh -c '"'"'echo ~ && sleep 0'"'"''
fatal: [consul2]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"
}
<consul1> (0, '/home/ansible\n', '')
<consul1> ESTABLISH SSH CONNECTION FOR USER: ansible
<consul1> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o 'IdentityFile="/home/myusername/.ssh/ansible_rsa"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ansible -o ConnectTimeout=10 -o ControlPath=/home/myusername/.ansible/cp/%h-%r consul1 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo /home/ansible/.ansible/tmp/ansible-tmp-1493659945.31-121706573316970 `" && echo ansible-tmp-1493659945.31-121706573316970="` echo /home/ansible/.ansible/tmp/ansible-tmp-1493659945.31-121706573316970 `" ) && sleep 0'"'"''
fatal: [consul3]: FAILED! => {
    "changed": false, 
    "failed": true, 
    "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"
}
<consul1> (0, 'ansible-tmp-1493659945.31-121706573316970=/home/ansible/.ansible/tmp/ansible-tmp-1493659945.31-121706573316970\n', '')
Using module file /usr/lib/python2.7/dist-packages/ansible/modules/files/stat.py
<consul1> ESTABLISH SSH CONNECTION FOR USER: ansible
<consul1> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o 'IdentityFile="/home/myusername/.ssh/ansible_rsa"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ansible -o ConnectTimeout=10 -o ControlPath=/home/myusername/.ansible/cp/%h-%r consul1 '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-lmgikasaenmhpipsihtiykkmysqwvnua; /usr/bin/python'"'"'"'"'"'"'"'"' && sleep 0'"'"''
<consul1> (0, '\n{"invocation": {"module_args": {"checksum_algorithm": "sha1", "get_checksum": true, "path": "/etc/consul.d/config.json", "checksum_algo": "sha1", "follow": false, "get_md5": false, "get_mime": true, "get_attributes": true}}, "stat": {"charset": "us-ascii", "uid": 0, "exists": true, "attr_flags": "e", "woth": false, "isreg": true, "device_type": 0, "mtime": 1493659695.2534463, "block_size": 4096, "inode": 133216, "isgid": false, "size": 467, "executable": false, "isuid": false, "readable": true, "version": "1178312012", "pw_name": "root", "gid": 0, "ischr": false, "wusr": true, "writeable": true, "mimetype": "text/plain", "blocks": 8, "xoth": false, "islnk": false, "nlink": 1, "issock": false, "rgrp": true, "gr_name": "root", "path": "/etc/consul.d/config.json", "xusr": false, "atime": 1493659696.5574582, "isdir": false, "ctime": 1493659695.5134487, "isblk": false, "wgrp": false, "checksum": "8edbed9e44ec14a840140b3b3fb08af33a441493", "dev": 64769, "roth": true, "isfifo": false, "mode": "0644", "xgrp": false, "rusr": true, "attributes": ["extents"]}, "changed": false}\n', '')
Using module file /usr/lib/python2.7/dist-packages/ansible/modules/files/file.py
<consul1> ESTABLISH SSH CONNECTION FOR USER: ansible
<consul1> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o 'IdentityFile="/home/myusername/.ssh/ansible_rsa"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ansible -o ConnectTimeout=10 -o ControlPath=/home/myusername/.ansible/cp/%h-%r consul1 '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-fzrbvdtymcgrpmroumbmbhczxxutxejp; /usr/bin/python'"'"'"'"'"'"'"'"' && sleep 0'"'"''
<consul1> (0, '\n{"group": "root", "uid": 0, "changed": false, "owner": "root", "state": "file", "gid": 0, "mode": "0644", "path": "/etc/consul.d/config.json", "invocation": {"module_args": {"directory_mode": null, "force": false, "remote_src": null, "path": "/etc/consul.d/config.json", "owner": "root", "follow": true, "group": "root", "unsafe_writes": null, "serole": null, "content": null, "state": null, "diff_peek": null, "setype": null, "dest": "/etc/consul.d/config.json", "selevel": null, "original_basename": "config.json.j2", "regexp": null, "validate": null, "src": null, "seuser": null, "recurse": false, "delimiter": null, "mode": "u=rw,g=r,o=r", "attributes": null, "backup": null}}, "diff": {"after": {"path": "/etc/consul.d/config.json"}, "before": {"path": "/etc/consul.d/config.json"}}, "size": 467}\n', '')
<consul1> ESTABLISH SSH CONNECTION FOR USER: ansible
<consul1> SSH: EXEC ssh -C -o ControlMaster=auto -o ControlPersist=60s -o StrictHostKeyChecking=no -o 'IdentityFile="/home/myusername/.ssh/ansible_rsa"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ansible -o ConnectTimeout=10 -o ControlPath=/home/myusername/.ansible/cp/%h-%r consul1 '/bin/sh -c '"'"'rm -f -r /home/ansible/.ansible/tmp/ansible-tmp-1493659945.31-121706573316970/ > /dev/null 2>&1 && sleep 0'"'"''
<consul1> (0, '', '')
ok: [consul1] => {
    "changed": false, 
    "diff": {
        "after": {
            "path": "/etc/consul.d/config.json"
        }, 
        "before": {
            "path": "/etc/consul.d/config.json"
        }
    }, 
    "gid": 0, 
    "group": "root", 
    "invocation": {
        "module_args": {
            "attributes": null, 
            "backup": null, 
            "content": null, 
            "delimiter": null, 
            "dest": "/etc/consul.d/config.json", 
            "diff_peek": null, 
            "directory_mode": null, 
            "follow": true, 
            "force": false, 
            "group": "root", 
            "mode": "u=rw,g=r,o=r", 
            "original_basename": "config.json.j2", 
            "owner": "root", 
            "path": "/etc/consul.d/config.json", 
            "recurse": false, 
            "regexp": null, 
            "remote_src": null, 
            "selevel": null, 
            "serole": null, 
            "setype": null, 
            "seuser": null, 
            "src": null, 
            "state": null, 
            "unsafe_writes": null, 
            "validate": null
        }
    }, 
    "mode": "0644", 
    "owner": "root", 
    "path": "/etc/consul.d/config.json", 
    "size": 467, 
    "state": "file", 
    "uid": 0
}
```
What's a bit curious on this is that it seems Ansible only connects to consul1 via SSH. Not consul2 or consul3.  Do you have that too in your environment? But I see no ``delegate_to`` at or near the task.
On other tasks the connection looks okay regarding the debug output.

**Comment by mrlesmithjr** [comment]
man I am confused on what could be your issue. I don't see this same result in any environment that I have tested this with so far. What Ansible version did you say you were using? I know I had ran into an issue with a 2.2 release but 2.3.0.0 seems fine. Do you have any hosts listed for the `consul_clients_group` or still `[]`?

Try creating a group in inventory (Without any hosts)
```
[consul_clients]
```
Then configure the following variable:
```
consul_clients_group: 'consul_clients'
```
Test that and report back. If that works I will work on changing some logic to not always account for clients but that should be a very rare instance.

**Comment by mrlesmithjr** [comment]
@jardleex Does your template show the following code block in `templates/etc/consul.d/config.json.j2` under the consul role?

```
{%   if inventory_hostname != consul_master %}
{%     set _temp_retry_join = [] %}
{%       for host in groups[consul_servers_group] %}
{%         if inventory_hostname != host and host in play_hosts %}
{%           set _temp_ip = hostvars[host]['consul_bind_address'] %}
{%           set _temp_retry_join = _temp_retry_join.append(_temp_ip) %}
{%         endif %}
{%       endfor %}
{%     set _config = config_json.update({"retry_join": _temp_retry_join}) %}
{%   endif %}
```

**Comment by mrlesmithjr** [comment]
@jardleex FYI...I have merged the code in from the `dev/retry-join-issue#14` branch via [PR#16]. This is now in `master`. I have validated from my testing that it works without even defining anything and leaving the default:
```
# Defines Ansible group that nodes belong to which are clients
# (agents or services running on)
consul_clients_group: 'consul_clients'
```
It works as designed now and configures the hosts which are the servers:
```
# Defines the Ansible group which contains the consul server(s)
consul_servers_group: 'consul_servers'
```
My static inventory:
```
# Generated by Vagrant

node0 ansible_ssh_host=127.0.0.1 ansible_ssh_port=2222 ansible_ssh_user='vagrant' ansible_ssh_private_key_file='/Users/larry/projects/vagrant/vagrant-box-templates/Ubuntu/xenial64/server/.vagrant/machines/node0/virtualbox/private_key'
node1 ansible_ssh_host=127.0.0.1 ansible_ssh_port=2200 ansible_ssh_user='vagrant' ansible_ssh_private_key_file='/Users/larry/projects/vagrant/vagrant-box-templates/Ubuntu/xenial64/server/.vagrant/machines/node1/virtualbox/private_key'
node2 ansible_ssh_host=127.0.0.1 ansible_ssh_port=2201 ansible_ssh_user='vagrant' ansible_ssh_private_key_file='/Users/larry/projects/vagrant/vagrant-box-templates/Ubuntu/xenial64/server/.vagrant/machines/node2/virtualbox/private_key'

[consul_servers]
node0
node1
node2
```

Results in no errors using [Ansible] `2.3.0` and ends up with the following configs:
```
vagrant@node0:~$ cat /etc/consul.d/config.json
{
    "acl_datacenter": "dc1",
    "acl_default_policy": "allow",
    "acl_down_policy": "extend-cache",
    "acl_master_token": "475be71a-03f1-4618-9dc5-07a8a739f6fa",
    "bind_addr": "192.168.250.10",
    "bootstrap": true,
    "client_addr": "0.0.0.0",
    "data_dir": "/var/consul",
    "datacenter": "dc1",
    "enable_syslog": true,
    "encrypt": "Se0NGFxRXObr+SFveE/c4g==",
    "log_level": "INFO",
    "node_name": "node0",
    "server": true
}
```
```
vagrant@node1:~$ cat /etc/consul.d/config.json
{
    "acl_datacenter": "dc1",
    "acl_default_policy": "allow",
    "acl_down_policy": "extend-cache",
    "acl_master_token": "475be71a-03f1-4618-9dc5-07a8a739f6fa",
    "bind_addr": "192.168.250.11",
    "bootstrap_expect": 3,
    "client_addr": "0.0.0.0",
    "data_dir": "/var/consul",
    "datacenter": "dc1",
    "enable_syslog": true,
    "encrypt": "Se0NGFxRXObr+SFveE/c4g==",
    "log_level": "INFO",
    "node_name": "node1",
    "retry_join": [
        "192.168.250.10",
        "192.168.250.12"
    ],
    "server": true
}
```
```
cat vagrant@node2:~$ cat /etc/consul.d/config.json
{
    "acl_datacenter": "dc1",
    "acl_default_policy": "allow",
    "acl_down_policy": "extend-cache",
    "acl_master_token": "475be71a-03f1-4618-9dc5-07a8a739f6fa",
    "bind_addr": "192.168.250.12",
    "bootstrap_expect": 3,
    "client_addr": "0.0.0.0",
    "data_dir": "/var/consul",
    "datacenter": "dc1",
    "enable_syslog": true,
    "encrypt": "Se0NGFxRXObr+SFveE/c4g==",
    "log_level": "INFO",
    "node_name": "node2",
    "retry_join": [
        "192.168.250.10",
        "192.168.250.11"
    ],
    "server": true
}
```
[Ansible]: <https://www.ansible.com>
[PR#16]: <https://github.com/mrlesmithjr/ansible-consul/pull/16>

**Comment by jardleex** [comment]
Hey,

I'm running on Ansible version 2.3.0.0. On a 2.2 release the issue was the same.
The last tests ran with an non defined/existing ``consul_clients_group``. I'll create an empty group tomorrow in my static inventory as requested and test that.

Yes, I do have this block in ``templates/etc/consul.d/config.json.j2``
````
{%   if inventory_hostname != consul_master %}
{%     set _temp_retry_join = [] %}
{%       for host in groups[consul_servers_group] %}
{%         if inventory_hostname != host and host in play_hosts %}
{%           set _temp_ip = hostvars[host]['consul_bind_address'] %}
{%           set _temp_retry_join = _temp_retry_join.append(_temp_ip) %}
{%         endif %}
{%       endfor %}
{%     set _config = config_json.update({"retry_join": _temp_retry_join}) %}
{%   endif %}
````

Thanks for merging PR#16 in.
Tomorrow I will try to deploy the Vagrant setup available within this repo.
Will let you know how it worked out.

Thanks again!

**Comment by mrlesmithjr** [comment]
Sounds good. I have updated the `Vagrant` folder with all of the most current roles for your testing.

**Comment by jardleex** [comment]
Hi,

just tested this:

```
[jardleex_servers]
consul1
consul2
consul3


[consul_clients]


---
- name: SIT.consul
  hosts: jardleex_servers
  vars:
    consul_servers_group: 'jardleex_servers'
    consul_clients_group: 'consul_clients'
  roles:
    - { role: mrlesmithjr.consul }
```
Result: **stil broken**


###
Tested the latest Vagrant box within the repo and guess what:
Result: **still broken**

```
....
TASK [ansible-consul : config | configuring consul] ****************************
fatal: [node1]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
fatal: [node3]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
fatal: [node4]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
fatal: [node2]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
ok: [node0]
....

```

Tested Vagrant box after commit ``5e00ebd9eb998d70dfc1dc948d7f8782df2aa8b3`` from Apr 28th:
Result: **still broken**
```
...
TASK [ansible-consul : config_consul | configuring consul] *********************
fatal: [node3]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
fatal: [node1]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
fatal: [node4]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
fatal: [node2]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
changed: [node0]
...
```

Tested Vagrant box after commit ``419bf79ca64deaa9c4970a4932e99a112e5c5615`` from Apr 23rd:
Result: **still broken**

```
...
TASK [ansible-consul : config_consul | configuring consul] *********************
fatal: [node1]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
fatal: [node3]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
fatal: [node4]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
fatal: [node2]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
changed: [node0]
...
```

More informations on my environment:
- OS of control machine: Ubuntu 16.04.2
- Kernel: 4.4.0-72-generic
- Python version: 2.7.12
- Python-jinja2: 2.8.1
- Vagrant: 1.9.4
- VirtualBox: 5.1.22-115126~Ubuntu~xenial

If it helps you, I can provide a list of all locally installed python packages with their version: As it is quite long I ain't just added it.

What OS, kernel and Python version are you using?

**Comment by jardleex** [comment]
Hi,

just tested the latest Vagrant box on a colleagues notebook:
Result: **broken**

Environment informations:
VirtualBox: 5.1.22
Python: 2.7.13
Python-jinja2: 2.8.1
Ansible: 2.2.1.0
OS: Debian 9
Kernel: 4.9.7



```
...
TASK [ansible-consul : config | configuring consul] ****************************
fatal: [node1]: FAILED! => {"censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result"}
fatal: [node2]: FAILED! => {"censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result"}
fatal: [node4]: FAILED! => {"censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result"}
fatal: [node3]: FAILED! => {"censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result"}
changed: [node0]
...
```

**Comment by mrlesmithjr** [comment]
And I just spun up the latest `Vagrant` without any issues. So this is definitely environmental specific.
```
PLAY RECAP *********************************************************************
node0                      : ok=36   changed=22   unreachable=0    failed=0
node1                      : ok=32   changed=20   unreachable=0    failed=0
node2                      : ok=32   changed=20   unreachable=0    failed=0
node3                      : ok=69   changed=47   unreachable=0    failed=0
node4                      : ok=69   changed=47   unreachable=0    failed=0
```

VirtualBox:
```
5.1.8r111374
```
OS:
```
ProductName:	Mac OS X
ProductVersion:	10.12.4
BuildVersion:	16E195
```
Python Virtualenv:
```
ansible==2.3.0.0
ansible-lint==3.4.12
appdirs==1.4.3
asn1crypto==0.22.0
astroid==1.5.2
backports.functools-lru-cache==1.3
cffi==1.10.0
configparser==3.5.0
cryptography==1.8.1
enum34==1.1.6
idna==2.5
ipaddress==1.0.18
isort==4.2.5
Jinja2==2.9.6
jmespath==0.9.2
lazy-object-proxy==1.2.2
MarkupSafe==1.0
mccabe==0.6.1
packaging==16.8
paramiko==2.1.2
pyasn1==0.2.3
pycparser==2.17
pycrypto==2.6.1
pylint==1.7.1
pyparsing==2.2.0
PyYAML==3.12
singledispatch==3.4.0.3
six==1.10.0
wrapt==1.10.10
```

And guess what. I just found the same error using `ansible 2.2.1.0`:
```
TASK [ansible-consul : config | configuring consul] ****************************
fatal: [node1]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
fatal: [node3]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
fatal: [node2]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
fatal: [node4]: FAILED! => {"changed": false, "failed": true, "msg": "AnsibleUndefinedVariable: 'None' has no attribute 'append'"}
ok: [node0]
```
So, I can confirm that on my end that `ansible 2.3.0.0` definitely works. I will do some regression testing to see which versions definitely work or not.

**Comment by mrlesmithjr** [comment]
`ansible-2.1.2.0` works as well

**Comment by mrlesmithjr** [comment]
I would also recommend maybe this role https://github.com/mrlesmithjr/ansible-control-machine or even this one https://github.com/mrlesmithjr/docker-ansible-toolbox

**Comment by mrlesmithjr** [comment]
I also use this script for setting up my virtualenvs for testing:
```
#!/usr/bin/env bash
set -x

ANSIBLE_VERSIONS=("1.9.4" "1.9.5" "1.9.6" "2.0.0.0" "2.0.0.1" "2.0.0.2" \
                 "2.0.1.0" "2.0.2.0" "2.1.0.0" "2.1.1.0" "2.1.2.0" \
                 "2.1.3.0" "2.1.4.0" "2.2.0.0" "2.2.1.0" "2.2.2.0" "2.3.0.0")
VIRTUALENV_PATH="./python-virtualenvs"

for ANSVER in "${ANSIBLE_VERSIONS[@]}"
do
  virtualenv $VIRTUALENV_PATH/ansible-$ANSVER
  source $VIRTUALENV_PATH/ansible-$ANSVER/bin/activate
  pip install ansible==$ANSVER ansible-lint
  deactivate
done
```

**Comment by jardleex** [comment]
Looks like I found the decayed tooth > **python-jinja2**!

Version **2.8.1**, which is default in Ubuntu 16.04 and Debian 9, is to old or buggy to fully support all features used in ``templates/etc/consul.d/config.json.j2``

I've downloaded the version **2.9.5** from [Ubuntu 17.04](http://packages.ubuntu.com/zesty/all/python-jinja2/download) and voila, the role is working!

I assume that your virtualenv setup always fetches the latest module versions so it always worked for you. As virtualenv nor docker are options for us we're using the OS default packages.

**Comment by mrlesmithjr** [comment]
Great catch. So, now to figure out the workaround. The virtualenvs are only setup on the control machine for Ansible. So, you could just use pip to install `ansible` and `python-jinja2` instead of using virtualenvs. I generally only use pip to install my python modules to ensure they are always the latest of pinned to a version.

**Comment by jardleex** [comment]
I've played a bit around with the for loop inside ``templates/etc/consul.d/config.json.j2`` but as already said my Jinja2 and Python voodoo is not very good so I ain't found another way of doing it.

Tomorrow I'll try to find a collegue which might help me on that.
Besides that I've began to work on a PR to simplify the bootstrap process. This won't solve the issue just make the template easier while removing config lines from older Consul versions.

**Comment by jardleex** [comment]
**FIX FOUND**

Hi @mrlesmithjr,

together with one of our DEV's we found a fix for this.
In ``templates/etc/consul.d/config.json.j2``
```
{%       set _ = _temp_retry_join.append(_temp_ip) %}
```
We successful tested this with python-jinja2 2.8.1 and 2.9.5 on Ubuntu 16.04 and Arch Linux.
Found the solution [here](https://groups.google.com/forum/#!topic/ansible-project/aiBiuUGWZGsl), last answer from _Olve Hansen_.

I'm already preparing a PR for this.
Please verify on your side when the PR comes in

Also please not that the role will mark ``config.json`` as changed on each run when a dynamic inventory is used. As Ansible requires a dict as inventory, the order of the nodes may differ. This leads to a different order for the _retry_join_ list which leads to a change.

I'll do two things on that:
1. Check if consul properly handles the _retry_join_ list even if the local node is included
2. If this fails I'll sort our dict in our dynamic inventory script before passing it to Ansible.

Thank you very much for your help and patience. I'm already working on a bunch of PR's to pimp this role even further!

Best regards

Jard

**Comment by mrlesmithjr** [comment]
Awesome...Thanks and glad you sorted it out. I have merged PR #20
